### ISSUE TITLE
Vulkan on Metal via Vulkan Portability

### ISSUE THREAD

**Issue by kvark** [DiligentGraphics/DiligentCore#69]
I've been testing the tutorial samples with gfx-portability, and with https://github.com/gfx-rs/portability/pull/177 resolved we have everything but the geometry and tessellation samples running on MacOS.

Would you consider changing the logic to use a generic Vulkan Portability library instead of MoltenVK specifically? Currently, it looks like Diligent links directly against `libMoltenVK.dylib`.

**Comment by kvark** [comment]
Also related to #46, #34, and #49

**Comment by DiligentGraphics** [comment]
Yes, that should be feasible. I think providing the path to the library during CMake configuration should work.
In fact in this case there will be no need to keep the binary in the repository. For macos, the size of libMoltenVK.so is about 2 MB, but for iOS it is almost 10 MB, so keeping it around is impossible withou bloating the repo size.

This will require little bit more effort from the user, but should simplify many things. Few default locations could be searched too I guess.

**Comment by kvark** [comment]
> I think providing the path to the library during CMake configuration should work.

A recommended way for Vulkan applications is to link to the loader (`libvulkan.1.dylib`) instead of the particular driver/implementation. Whether you want to bundle a particular version of the loader with an app is one question (that isn't important to us), and I think you already do, but not link to it (at least with default build settings). The implementation is then selected with the ICD JSON file, path to which can be easily overridden by the user with `VK_ICD_FILENAMES` environment variable (handled by the loader).

Also, going forward (i.e. this isn't required now), there is going to be a few Vulkan Portability extensions that work across gfx-portability and MoltenVK:
  - [VK_EXTX_portability_subset](https://github.com/KhronosGroup/Vulkan-Portability/blob/53be040f04ce55463d0e5b25fd132f45f003e903/doc/specs/VK_EXTX_portability_subset.md): to discover the extra limitations
  - window system integration: to create a surface from `CAMetalLayer`
  - direct `MTLTexture` and `MTLDevice` interop: for video decoding and such

**Comment by DiligentGraphics** [comment]
@kvark I am correct that if Diligent links against libvulkan.1.dylib, the loader will take care of setting up all the function pointers being it MoltenVk or any other implementation? Is libvulkan.1.dylib part of the LunarG SDK?

**Comment by kvark** [comment]
Yes

**Comment by DiligentGraphics** [comment]
OK, will do.

**Comment by kvark** [comment]
@DiligentGraphics note that if you do any testing on macOS and encounter issues with MoltenVK, and you link with the loader, you can try running on gfx-portability if you provide `VK_ICD_FILENAMES=<path_to_portability_json>` in the environment. If the behavior is different, it might be a bug in either of the libraries, and we'd be interested to fix it.

**Comment by DiligentGraphics** [comment]
Ok, I am actually looking into this right now. So to make it work with MoltenVk, I should have a loader and a json file next to it? How will it find the implementation? Is it also soecified by json? 
Also, how does this work on iOS. It seems there is no loader for iOS in Vulkan SDK.

**Comment by kvark** [comment]
The [loader workflow](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/master/loader/LoaderAndLayerInterface.md) is the same as on other platforms:
  - you link to the loader dylib
  - when an app starts, the loader searches for available ICD jsons in some folders, plus whatever you provide in `VK_ICD_FILENAMES` environment
  - the ICD json points to an actual driver/portability dynamic library

gfx-portability can be either downloaded from [our releases](https://github.com/gfx-rs/portability/releases) or built from source (via `make`, there are variations in what you can build of course). In either case the [sample JSON](https://github.com/gfx-rs/portability/blob/57796417387c61bf19841c39b7c903b36d8b0a96/libportability-icd/portability-macos-debug.json) is provided.

As for iOS, I don't really know. You can always link to the portability library directly instead of the loader though. In our case, [libportability](https://github.com/gfx-rs/portability/tree/master/libportability) is the one exposing driver entry points, while [libportability-icd](https://github.com/gfx-rs/portability/tree/master/libportability-icd) exposes only ICD entry points.

**Comment by DiligentGraphics** [comment]
@kvark I tried to link against libvulkan.1.dylib from LunarG SDK and that seems to work fine. However, it wants libMoltenVK.dylib to be located in the same folder otherwise it fails to link.
I can't understand where dependency on MoltenVK comes from.

otool -L libvulkan.1.dylib prints this:

```
@rpath/libvulkan.1.dylib (compatibility version 1.0.0, current version 1.1.101)
/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.200.5)
/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1560.12.0)
```

**Comment by kvark** [comment]
@DiligentGraphics the loader links to neither of the drivers *dynamically*, so `otool` will not show you it, this is expected. Instead, the loader scans for ICD jsons. I see that it bundles the `MoltenVK_icd.json` in `sdk/macOS/etc/vulkan/icd.d/`. I also see that sample applications in the SDK carry that ICD inside, i.e. `sdk/Applications/vkcube.app/Contents/Resources/vulkan/icd.d/MoltenVK_icd.json`.

TL;DR; if you want to run with a specific portability library during development, just provide the `VK_ICD_FILENAMES=<path_to_icd_json>` in the environment. Perhaps, for shipping you just make sure the corresponding ICD is bundled in `Contents/Resources/vulkan/` and distribute the app with both `libvulkan.1.dylib` and the portability library. Not completely sure, tbh.

**Comment by DiligentGraphics** [comment]
@kvark I still don't understand why the build fails. If the whole point is to be able to switch Vulkan implementations dynamically at run time, then why does it want specific implementation (MoltenVK) at build time? If the loader searches for specific dylib based on json or any other configuration, shouldn't I have run-time error if it can't find the library instead of build error?

**Comment by kvark** [comment]
Wait, you didn't say that it was the *build* that failed. This makes no sense. Perhaps it's your project settings that still expect `libMoltenVK.dylib` somewhere? `libvulkan.1.dylib` is definitely not what fails for you.

On an unrelated note, we've just made a release that mentioned your engine: https://github.com/gfx-rs/portability/releases/tag/0.5

**Comment by DiligentGraphics** [comment]
Ah, darn it! Forgot that CMake does not update the library path once it found it unless you delete that CMake variable... Now everything builds fine - almost there.


> On an unrelated note, we've just made a release that mentioned your engine: https://github.com/gfx-rs/portability/releases/tag/0.5

That is really cool! Thanks!

**Comment by DiligentGraphics** [comment]
@kvark I tried to use gfx-portability instead of MoltenVK, but it fails with the following errors:

```
Diligent Engine: ERROR: Vulkan debug message (ERROR, general): Message Id: 0 | Message Name: Loader Message
                 terminator_CreateDevice: Failed in ICD .../gfx-portability-macos-0.5/libportability-icd/../target/debug/libportability_icd.dylib vkCreateDevicecall
                 Object[0] (instance): Handle 0x1040f1200

Diligent Engine: ERROR: Vulkan debug message (ERROR, general): Message Id: 0 | Message Name: Loader Message
                 vkCreateDevice:  Failed to create device chain.
                 Object[0] (instance): Handle 0x1040f1200

Diligent Engine: CRITICAL ERROR in VulkanLogicalDevice() (VulkanLogicalDevice.cpp, 51): Failed to create logical device
VK Error Code: ERROR_FEATURE_NOT_PRESENT
```

Do you know what may be wrong here?

**Comment by kvark** [comment]
It says a feature does not present. Do you discover physical device features before creating the device? We can work on exposing a feature you need, just need to know which one.

**Comment by DiligentGraphics** [comment]
Ah, OK. Yeah, being more explicit about what feature exactly is missing would be really helpful here. The feature was `multiViewport`. None of the samples/tutorials really need this feature, but the third-party UI library does. I already had issues with this and even disabled it once, but then enabled again. I guess it is time to disable it for good https://github.com/DiligentGraphics/DiligentSamples/issues/8.

Anyway, when this feature is not enabled, the device is created successfully and all tutorials run fine 
(yay!) with one exception: I recently made some updates to avoid GPU falling indefinitely behind the CPU (https://github.com/KhronosGroup/MoltenVK/issues/530): https://github.com/DiligentGraphics/DiligentCore/commit/b7cbf119119eeeee4026d6f631c7de3b65fa929d. With this change, [this call](https://github.com/DiligentGraphics/DiligentCore/blob/master/Graphics/GraphicsEngineVulkan/src/SwapChainVkImpl.cpp#L450)

```cpp
LogicalDevice.WaitForFences(1, &OldestSubmittedFence, VK_TRUE, UINT64_MAX);
```

returns `VK_TIMEOUT`  which does not make sense as timeout is set to `UINT64_MAX`. This causes debug validation error in the next line. This works fine on other platforms/implementations.

**Comment by kvark** [comment]
@DiligentGraphics this is great news!

> being more explicit about what feature exactly is missing would be really helpful here

In general, I agree. But it's hard to pinpoint the party that should be responsible for this:
  1. an application can do, so why should other layers bother?
  2. validation layer can do, but this isn't really an invalid usage: there can be app that properly reacts on that error
  3. driver/portability layer can do

In general, we spew out warnings/infos/errors using standard Rust mechanisms. gfx-portability can be built (with `--features env_logger`) to output these messages to the screen. Currently though, we don't have any message for features missing, I'll add it.

> returns VK_TIMEOUT which does not make sense as timeout is set to UINT64_MAX

I double-checked the code and there is only one way this could be happening: if the fence is not pending (i.e. not scheduled), and it was already reset. Any specified timeout then doesn't make sense. Interestingly, I don't see anything in the spec defining the behavior in this case, so I guess you are stepping into undefined behavior? I'd at least expect this in some way mentioned in valid use statements.

btw, the spec says you don't need to get a fence status before waiting:
> If the condition is satisfied when vkWaitForFences is called, then vkWaitForFences returns immediately. 

Anyway, I've made a patch to gfx-rs to log both of these warnings - https://github.com/gfx-rs/gfx/pull/2708. We'll also make the published gfx-portability debug builds to enable output to terminal.

**Comment by DiligentGraphics** [comment]
> I double-checked the code and there is only one way this could be happening: if the fence is not pending (i.e. not scheduled), and it was already reset.

There is special logic [here](https://github.com/DiligentGraphics/DiligentCore/blob/master/Graphics/GraphicsEngineVulkan/src/SwapChainVkImpl.cpp#L445), [here](https://github.com/DiligentGraphics/DiligentCore/blob/master/Graphics/GraphicsEngineVulkan/src/SwapChainVkImpl.cpp#L454), and [here](https://github.com/DiligentGraphics/DiligentCore/blob/master/Graphics/GraphicsEngineVulkan/src/SwapChainVkImpl.cpp#L460) that assures that the fence status is only checked if it has been submitted. In fact there were few issues causing deadlock on Windows when checking the status of the fence that has not been submitted. Everything works fine now on other platforms without any validation layer messages. Now that there is proper loader on Mac, validation layers can also be enabled. On MoltenVk, all samples and tutorials run clear. With gfx-portability, there is a lot of messages, including the one about using fence that has not been submitted:

```
Diligent Engine: ERROR: Vulkan debug message (validation): VUID-vkResetFences-pFences-01123
                 Fence 0x5e is in use. The Vulkan spec states: Each element of pFences must not be currently associated with any queue command that has not yet completed execution on that queue (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkResetFences-pFences-01123)
                 Object[0] (fence): Handle 0x5e
```

As I mentioned the fact that it runs clear on other implementations including MoltenVk most likely suggests there is an issue in gfx-portability.

There are also some messages that make no sense to me:

```
Diligent Engine: ERROR: Vulkan debug message (validation): VUID-VkViewport-x-01232
                 vkCmdSetViewport: pViewports[0].x + pViewports[0].width (=0.000000 + 1600.000000 = 1600.000000) is greater than VkPhysicalDeviceLimits::viewportBoundsRange[1] (=0.000000). The Vulkan spec states: (x + width) must be less than or equal to viewportBoundsRange[1] (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkViewport-x-01232)
                 Object[0] (cmd buffer): Handle 0x10405f000

Diligent Engine: ERROR: Vulkan debug message (validation): VUID-VkViewport-y-01776
                 vkCmdSetViewport: pViewports[0].y (=1200.000000) exceeds VkPhysicalDeviceLimits::viewportBoundsRange[1] (=0.000000). The Vulkan spec states: y must be less than or equal to viewportBoundsRange[1] (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkViewport-y-01776)
                 Object[0] (cmd buffer): Handle 0x10405f000


Diligent Engine: ERROR: Vulkan debug message (validation): VUID-VkPipelineLayoutCreateInfo-setLayoutCount-00286
                 vkCreatePipelineLayout(): setLayoutCount (2) exceeds physical device maxBoundDescriptorSets limit (0). The Vulkan spec states: setLayoutCount must be less than or equal to VkPhysicalDeviceLimits::maxBoundDescriptorSets (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkPipelineLayoutCreateInfo-setLayoutCount-00286)
                 Object[0] (unknown): Handle 0x0

Diligent Engine: ERROR: Vulkan debug message (validation): VUID-VkPipelineLayoutCreateInfo-pSetLayouts-00287
                 vkCreatePipelineLayout(): max per-stage sampler bindings count (1) exceeds device maxPerStageDescriptorSamplers limit (0). The Vulkan spec states: The total number of descriptors of the type VK_DESCRIPTOR_TYPE_SAMPLER and VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER accessible to any shader stage across all elements of pSetLayouts must be less than or equal to VkPhysicalDeviceLimits::maxPerStageDescriptorSamplers (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkPipelineLayoutCreateInfo-pSetLayouts-00287)
                 Object[0] (unknown): Handle 0x0
```
and many more. I guess you can now easily enable validation layers and switch implementations to try it yourself. I also fixed sample apps and they do not enable multiViewport feature anymore.

**Comment by kvark** [comment]
@DiligentGraphics thank you for the info, it's great feedback for us!
I debugged your engine a bit and found a problem in our implementation: we currently don't expect both semaphore and fence to be provided. Filed https://github.com/gfx-rs/gfx/issues/2711 to get that sorted out.

> It also looks like deviceLimits structure is not initialized that causes a lot of messages like these:

Yeah, we are slowly expanding the meaningful limits that are exposed. We'll get those filled out ;)

**Comment by DiligentGraphics** [comment]
@kvark I also have a test application that is essentially a collection of unit tests. They are not the most elegant unit tests you will be able to find out there, but they cover pretty much the entire Diligent API and implementation. I keep hoping to rework them one day in a true test system that runs as part of CI, but realistically not sure when this may happen. Anyway, if this app runs on a platform, I have pretty high confidence that there are no major problems. There are currently few issues on MacOS, but if you think it may be useful for you, I can fix it.

![image](https://user-images.githubusercontent.com/15135271/54860588-98ff6780-4cd9-11e9-808f-4787b1cd86c4.png)

**Comment by kvark** [comment]
Yes, this would definitely be useful to us. There are many ways we currently apply for testing, including Vulkan CTS, but neither method is best. Having your demo as an option would be helpful.

**Comment by DiligentGraphics** [comment]
OK, will fix it.

**Comment by kvark** [comment]
FYI, the fixes to the issues you found are coming in https://github.com/gfx-rs/portability/pull/180

**Comment by DiligentGraphics** [comment]
I am getting this error when building gfx-portability:

```
   Compiling portability-gfx v0.1.0 (/git/portability/libportability-gfx)
error[E0463]: can't find crate for `back`
  --> libportability-gfx/src/lib.rs:15:1
   |
15 | extern crate gfx_backend_vulkan as back;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: Could not compile `portability-gfx`.
```

Is there a way to download pre-built binaries?

**Comment by kvark** [comment]
@DiligentGraphics all you need to build is `make`, then the (debug) ICD library appears in `target/debug/libportability_icd.dylib`. For release, you'd do `make release` correspondingly.

We have the CI infrastructure to generate binaries on tags/releases, but we've just make 0.5 and it's too early to release an update.

**Comment by DiligentGraphics** [comment]
@kvark With your patch, all tutorials run fine (though still generate a lot of validation messages). There also seems to be an issue with blend modes in Tutorial10 since it looks little different. 

What would be really cool if Atmosphere sample worked. It doesn't run on MoltenVK either (fails to convert one of the shaders with odd error about row-major matrices that are never used). gfx-portability in fact goes beyond that point and crashes at draw command:

```
-[MTLDebugRenderCommandEncoder validateFramebufferWithRenderPipelineState:]:1218: failed assertion `For color attachment 0, the render pipeline's pixelFormat (MTLPixelFormatRG8Unorm) does not match the framebuffer's pixelFormat (MTLPixelFormatRG32Float).'
```

It feels like it is checking PSO-Frame Buffer compatibility when PSO is bound, but proper Framebuffer is bound later. The sample works fine and runs clear on Windows & Linux.

**Comment by kvark** [comment]
@DiligentGraphics 
> With your patch, all tutorials run fine (though still generate a lot of validation messages)

Great, thanks for confirming! The validation errors in this case are unlikely going to help much. At best they'll tell us which limits are missing (not that we don't know that already....).

> There also seems to be an issue with blend modes in Tutorial10 since it looks little different.

Thanks for the pointer, I'll have a look!

> What would be really cool if Atmosphere sample worked. It doesn't run on MoltenVK either (fails to convert one of the shaders with odd error about row-major matrices that are never used).

Unfortunately, we also rely on SPIRV-Cross for shader compilation still. Our plan is to remove this dependency, but we don't have resources yet to develop an in-house alternative. This means that for now we are going to have the same shader errors as MoltenVK.

> It feels like it is checking PSO-Frame Buffer compatibility when PSO is bound, but proper Framebuffer is bound later. The sample works fine and runs clear on Windows & Linux.

Interesting! May I guess that your sequence of actions is:
  1. enter a pass A
  2. bind pipeline P that is not compatible with pass A
  3. leave pass A, enter pass B
  4. draw something with this pipeline P

Our implementation doesn't expect the fact that a pipeline is bound in a render pass that is incompatible, although I do agree this is technically valid Vulkan (if that pipeline isn't used in this pass). I can think of a way to solve this (similar to vertex buffers mechanics).

**Comment by DiligentGraphics** [comment]
> May I guess that your sequence of actions is

Correct. The engine does not enforce specific order for setting pipeline states and binding render targets (there are no native render passes and frame buffer objects in the common API yet, but they are planned). So a PSO may be temporarily incompatible with current framebuffer.

**Comment by kvark** [comment]
About Tutorial10. You said the blending looks different. Is it different from Vulkan? Or from MoltenVK? I noticed that BlendState array (on your side) isn't initialized to zero and only partially filled, so there is that.

**Comment by kvark** [comment]
I checked Tutorial10 and didn't find anything to fix, see https://github.com/gfx-rs/gfx/issues/2714#issuecomment-477178743

Also, we have a new release - https://github.com/gfx-rs/portability/releases/tag/0.6 . Binaries are going to be published there momentarily by CI. It carries the fixes to problems you pointed to here, including binding the incompatible pipeline in a pass. (note: Atmosphere sample now fails in shader compilation, as expected).

**Comment by DiligentGraphics** [comment]
This is screenshot with gfx-portability:

![image](https://user-images.githubusercontent.com/15135271/55094367-84bdc080-5073-11e9-9755-cd15c58e30c2.png)

This is what it is supposed to look like:

![image](https://user-images.githubusercontent.com/15135271/55094523-c77f9880-5073-11e9-965d-344a1af6a214.png)

It is not completely broken, but the look is definitely different.

**Comment by kvark** [comment]
Could it be just the swapchain format being different, i.e. sRGB versus regular Unorm?

**Comment by DiligentGraphics** [comment]
Unlikely. The background will be different as well as overall brightness. There are only some polygons that look different, and also all other tutorials look correct
